core/src/main/scala/kafka/server/BrokerToControllerChannelManager.scala

```diff
@@ -0,0 +1,211 @@
+        false,
+        new ApiVersions,
+        logContext
+      )
+    }
+    val threadName = threadNamePrefix match {
+      case None => s"broker-${config.brokerId}-to-controller-send-thread"
+      case Some(name) => s"$name:broker-${config.brokerId}-to-controller-send-thread"
+    }
+
+    new BrokerToControllerRequestThread(networkClient, manualMetadataUpdater, requestQueue, metadataCache, config,
+      brokerToControllerListenerName, time, threadName)
+  }
+
+  private[server] def sendRequest(request: RequestChannel.Request,
+                                  callback: RequestCompletionHandler): Unit = {
+    val requestBuilder = new NoOpRequestBuilder(request.context.header.apiKey, request.body[AbstractRequest])
+    requestQueue.put(BrokerToControllerQueueItem(requestBuilder, callback, ""))
+  }
+
+  private[server] def forwardRequest(originalRequest: RequestChannel.Request,
+                                     callback: RequestCompletionHandler): Unit = {
+    val requestBuilder = new NoOpRequestBuilder(originalRequest.context.header.apiKey, originalRequest.body[AbstractRequest])
+    requestQueue.put(BrokerToControllerQueueItem(requestBuilder, callback, originalRequest.context.principal.getName))
+  }
+}
+
+case class BrokerToControllerQueueItem(request: AbstractRequest.Builder[_ <: AbstractRequest],
+                                       callback: RequestCompletionHandler,
+                                       initialPrincipalName: String)
+
+class BrokerToControllerRequestThread(networkClient: KafkaClient,
+                                      metadataUpdater: ManualMetadataUpdater,
+                                      requestQueue: BlockingQueue[BrokerToControllerQueueItem],
+                                      metadataCache: kafka.server.MetadataCache,
+                                      config: KafkaConfig,
+                                      listenerName: ListenerName,
+                                      time: Time,
+                                      threadName: String)
+  extends InterBrokerSendThread(threadName, networkClient, time, isInterruptible = false) {
+
+  private var activeController: Option[Node] = None
+
+  override def requestTimeoutMs: Int = config.controllerSocketTimeoutMs
+
+  override def generateRequests(): Iterable[RequestAndCompletionHandler] = {
+    val requestsToSend = new mutable.Queue[RequestAndCompletionHandler]
+    if (requestQueue.peek() != null) {
+      val topRequest = requestQueue.take()
+
+      val request = RequestAndCompletionHandler(
+        activeController.get,
+        topRequest.request,
+        handleResponse(topRequest),
+        topRequest.initialPrincipalName)
+      requestsToSend.enqueue(request)
+    }
+    requestsToSend
+  }
+
+  private[server] def handleResponse(request: BrokerToControllerQueueItem)(response: ClientResponse): Unit = {
+    if (response.wasDisconnected()) {
+      activeController = None
+      requestQueue.put(request)
```

Comment:
If other requests are queued up, I think this could mess up the ordering.

If we have requests`[A, B, C, D]` in the `requestQueue`, and A popped off is sent out, and the controller channel is disconnected, we will then have a queue like `[B, C, D, A]`. I think we need strict ordering for controller requests, so we would need to handle this differently. 

Maybe if we used a deque for the request queue, we could fix this by putting a failed item at the head of the queue. That, or the `doWork` call would need to hold onto a queue item until it had succeeded or failed irrevocably. But, I would guess we shouldn't block too long in the doWork call.

Reply:
If active controller could not be detected, it also makes sense to block indefinitely I guess.

Reply:
Replaced with `LinkedBlockingDeque`